### DISCUSSION TITLE
A question refers to Full.ino example

### DISCUSSION THREAD

**Discussion by Sennopati** [contrem/arduino-timer#81] [Q&A]
Hi All,

Inside loop() function, I added "String str = Serial.readString()" as shown in the below quote which causes the LED blinking speed cannot be changed anymore about 1 second on and off. Any values put inside the following statement "timer.every(250, toggle_led);" do not impact the blinking speed. 

Serial.readString() is required to give instruction from another application to control the LED behavior according to the particular circumstance. Please explain how to solve this problem. thanks!

void loop() {
  String str = Serial.readString();
  
  timer.tick(); // tick the timer
  t_timer.tick();
  u_timer.tick();
}

**Comment by skrobinson** [comment]
What is your Serial timeout for `String str = Serial.readString()`? The default is 1000ms.

**Comment by philj404** [comment]
As @skrobinson suggests, [`Serial.readString()`](https://reference.arduino.cc/reference/en/language/functions/communication/serial/readstring/) blocks until it times out.  Nothing other than reading the string can happen while it is running.

You _could_ change the default timeout with [`Serial.setTimeout()`](https://reference.arduino.cc/reference/en/language/functions/communication/serial/settimeout/). `Serial.readString()` would then block for a different amount of time.  But note that if you set the timeout to say 200ms, the other timer tasks would only have 50ms remaining before they will be blocked again.  Any tasks that need to repeat faster than 200ms will run late.

Do you actually need to wait for that timeout?  Another approach would be to create a task that is more "cooperative" and avoid blocking.  Can you use [`Serial.available()`](https://reference.arduino.cc/reference/en/language/functions/communication/serial/available/) and [`Serial.read()`](https://reference.arduino.cc/reference/en/language/functions/communication/serial/read/) or [`Serial.readBytes()`](https://reference.arduino.cc/reference/en/language/functions/communication/serial/readbytes/) to process bytes as they become ready, and just check for more bytes later?

**Comment by philj404** [comment]
Also since `String str = Serial.readString();` executes every time `loop()` is called, it gets called a LOT.  That timeout is slowing down each loop.  It probably does not need to run that often -- particularly if it waits one second.

Can you create a task to process the `Serial.readString();` instead of putting it directly in `loop()`?  This gives the timer get a chance to find other things to do while waiting.

You may not need any timeout at all (set timeout to 0), if you can trust the timer task will read serial data often enough to keep the `Serial` buffers from overflowing.

For example:
```
bool processSerial(void *) {
    String str = Serial.readString();

    // do what is needed with str...

    return true;  // repeat
}
...

void setup() {
    // other setup...

    Serial.setTimeout(0); // make Serial cooperative

    // make processSerial run often enough to keep buffers from overflowing
    int serialTime = 1000; // about equivalent to a 1 second timeout

    timer.every(serialTime, processSerial); 
}
...

void loop() {
    // calling Serial.readString() in a task now

    timer.tick(); // tick the timer
    t_timer.tick();
    u_timer.tick();
}
```

**Comment by Sennopati** [comment]
thanks so much all!

I've written Serial.setTimeout(100) and it works.

On Wed, Mar 8, 2023 at 7:45 AM Phil Jansen ***@***.***> wrote:

> Also since String str = Serial.readString(); executes every time loop()
> is called, it gets called a LOT. That timeout is slowing down each loop. It
> probably does not need to run that often -- particularly if it waits one
> second.
>
> Can you create a task to process the Serial.readString(); instead of
> putting it directly in loop()? This gives the timer get a chance to find
> other things to do while waiting.
>
> You may not need any timeout at all (set timeout to 0), if you can trust
> the timer task will read serial data often enough to keep the Serial
> buffers from overflowing.
>
> For example:
>
> bool processSerial(void *) {
>     String str = Serial.readString();
>
>     // do what is needed with str...
>
>     return true;  // repeat
> }
> ...
>
> void setup() {
>     // other setup...
>
>     Serial.setTimeout(0); // make Serial cooperative
>
>     // make processSerial run often enough to keep buffers from overflowing
>     int serialTime = 1000; // about equivalent to a 1 second timeout
>
>     timer.every(serialTime, processSerial);
> }
> ...
>
> void loop() {
>     // calling Serial.readString() in a task now
>
>     timer.tick(); // tick the timer
>     t_timer.tick();
>     u_timer.tick();
> }
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/contrem/arduino-timer/discussions/81#discussioncomment-5235173>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A6D4F7A7NEUDIOHHZQKIV43W27JD3ANCNFSM6AAAAAAVH45I2I>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Sandy Wibowo

**Comment by philj404** [comment]
I am glad that changing the timeout works for you.  Just remember the way the code is set up that `Serial.readString()` will wait 100ms before returning. That means that each call of `loop()` will have a 100ms delay before it calls `timer.tick()`.

This may not be a problem for you.  

But if your sketch needs more time for task calculations, or you want the timers to have finer resolution than 100ms, you will need to shorten that timeout even more.

Have fun!
